src/Nethermind/Nethermind.State/StorageProvider.cs

```diff
@@ -38,26 +40,28 @@ public class StorageProvider : IStorageProvider
 
         private readonly ILogger _logger;
 
-        private readonly ISnapshotableDb _stateDb;
+        private readonly IKeyValueStore _stateDb;
         private readonly IStateProvider _stateProvider;
 
         private ResettableDictionary<Address, StorageTree> _storages = new ResettableDictionary<Address, StorageTree>();
 
         private const int StartCapacity = Resettable.StartCapacity;
         private int _capacity = StartCapacity;
-        private Change[] _changes = new Change[StartCapacity];
+        private Change?[] _changes = new Change[StartCapacity];
         private int _currentPosition = -1;
 
-        public StorageProvider(IDbProvider dbProvider, IStateProvider stateProvider, ILogManager logManager) 
-            :this(dbProvider.StateDb, stateProvider, logManager)
-        {
-        }
-        
-        public StorageProvider(ISnapshotableDb stateDb, IStateProvider stateProvider, ILogManager logManager)
+        public StorageProvider(
+            IKeyValueStore? stateDb,
+            IStateProvider? stateProvider,
+            IWitnessCollector? witnessCollector,
+            ILogManager? logManager)
         {
             _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
-            _stateDb = stateDb ?? throw new ArgumentNullException(nameof(stateDb));
             _stateProvider = stateProvider ?? throw new ArgumentNullException(nameof(stateProvider));
+            _stateDb = stateDb ?? throw new ArgumentNullException(nameof(stateDb));
+            _stateDb = witnessCollector == NullWitnessCollector.Instance
+                ? _stateDb
+                : new WitnessingStore(_stateDb, witnessCollector);
```

Comment:
I don't like design that we have one parameter just to modify 2nd parameter. That creates some weird coupling. Can we pass WItnessingStore here instrad of stateDb when we are using witnesses?

We could also add a decorator or derived class to do that parameter modification to decouple this a bit.

Reply:
yeah, obviously, this looks silly, thanks